subsys/dfu_data_writer/src/dfu_data_writer.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdint.h>
+#include <zephyr.h>
+#include <flash.h>
+#include <dfu/mcuboot.h>
+#include <pm_config.h>
+#include <logging/log.h>
+#include <dfu_data_writer.h>
+
+LOG_MODULE_REGISTER(dfu_data_writer, CONFIG_DFU_DATA_WRITER_LOG_LEVEL);
+
+static u32_t	flash_address;
+static struct	device *flash_dev;
+static bool	is_flash_page_erased[FLASH_PAGE_MAX_CNT];
+
+static int flash_page_erase_if_needed(u32_t address)
+{
+	struct flash_pages_info info;
+	int err = flash_get_page_info_by_offs(flash_dev, address, &info);
+
+	if (err != 0) {
+		LOG_ERR("flash_get_page_info_by_offs error %d\n", err);
+		return err;
+	}
+	if (!is_flash_page_erased[info.index]) {
+		err = flash_write_protection_set(flash_dev, false);
+		if (err != 0) {
+			LOG_ERR("flash_write_protection_set error %d\n", err);
+			return err;
+		}
+		err = flash_erase(flash_dev, info.start_offset, info.size);
+		if (err != 0) {
+			LOG_ERR("flash_erase error %d at address %08x\n",
+				err, info.start_offset);
+			return err;
+		}
+		is_flash_page_erased[info.index] = true;
+		err = flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			LOG_ERR("flash_write_protection_set error %d\n", err);
+			return err;
+		}
+	}
+	return 0;
+}
+
+int dfu_data_writer_verify_size(size_t size)
+{
+	if (size > PM_MCUBOOT_SECONDARY_SIZE) {
+		LOG_ERR("Requested file too big to fit in flash\n");
+		return -EFBIG;
+	}
+	return 0;
+}
+
+int dfu_data_writer_finalize(void)
```

Comment:
I've been thinking about @carlescufi's comment. If we move `dfu_data_writer_finalize` into the other module, and allow the area start and size to be passed to init, we'd have a much more generalized flash_area stream end point. We could alternatively rename _finalize to _mcuboot_tag or something, allowing the _init and _fragment can be used separately.

Reply:
There is a rather large overhead in having the `flash_page_erase_if_needed` shared. This library also contains a map of all deleted flash pages, which makes it non-trivial to split.

Reply:
Actually, such a library already exists, `zephyr/include/dfu/flash_img.h` but I recollect being told not to use this for performance reasons.

Reply:
Ok, then I propose changing the name to something mcuboot-specific.